tables/AlternateMans/BufferedColumnarFile.h

```diff
@@ -0,0 +1,307 @@
+      : RowBasedFile(filename, header_size, stride),
+        packed_buffer_((stride + 7) / 8),
+        rows_per_block_(std::max<size_t>(1, BufferSize / stride)),
+        block_buffer_(rows_per_block_ * stride) {}
+
+  // Open an existing columnar file
+  VarBufferedColumnarFile(const std::string& filename, size_t header_size)
+      : RowBasedFile(filename, header_size) {
+    if (Stride() != 0) {
+      packed_buffer_.resize((Stride() + 7) / 8);
+      rows_per_block_ = std::max<size_t>(1, BufferSize / Stride());
+      block_buffer_.resize(rows_per_block_ * Stride());
+    }
+    active_block_ = std::numeric_limits<uint64_t>::max();
+  }
+
+  void ActivateBlock(uint64_t row) {
+    const uint64_t block = row / rows_per_block_;
+    if (active_block_ != block) {
+      if (block_changed_) {
+        const uint64_t start_row = active_block_ * rows_per_block_;
+        const size_t n_rows_to_write =
+            std::min(rows_per_block_, std::max(NRows(), start_row) - start_row);
+        Seek(start_row * Stride() + DataLocation(), SEEK_SET);
+        WriteData(block_buffer_.data(), n_rows_to_write * Stride());
+        block_changed_ = false;
+      }
+
+      const uint64_t start_row = block * rows_per_block_;
+      const size_t n_rows_to_read =
+          std::min(rows_per_block_, std::max(NRows(), start_row) - start_row);
+      if (n_rows_to_read > 0) {
+        Seek(start_row * Stride() + DataLocation(), SEEK_SET);
+        ReadData(block_buffer_.data(), n_rows_to_read * Stride());
+      }
+      std::fill(block_buffer_.begin() + n_rows_to_read * Stride(),
+                block_buffer_.end(), 0);
+
+      active_block_ = block;
+    }
+  }
+
+  template <typename ValueType>
+  void ReadImplementation(uint64_t row, uint64_t column_offset, ValueType* data,
+                          uint64_t n) {
+    assert(column_offset + n * sizeof(ValueType) <= Stride());
+    if (row >= NRows()) {
+      std::fill_n(data, n, ValueType());
+    } else {
+      ActivateBlock(row);
+      const uint64_t block_row = active_block_ * rows_per_block_;
+      const unsigned char* position =
+          block_buffer_.data() + (row - block_row) * Stride() + column_offset;
+      std::copy_n(position, n * sizeof(ValueType),
+                  reinterpret_cast<unsigned char*>(data));
+    }
+  }
+
+  template <typename ValueType>
+  void WriteImplementation(uint64_t row, uint64_t column_offset,
+                           const ValueType* data, uint64_t n) {
+    assert(column_offset + n * sizeof(ValueType) <= Stride());
+    ActivateBlock(row);
+    const uint64_t block_row = active_block_ * rows_per_block_;
+    unsigned char* position =
+        block_buffer_.data() + (row - block_row) * Stride() + column_offset;
+    std::copy_n(reinterpret_cast<const unsigned char*>(data),
+                n * sizeof(ValueType), position);
+    SetNRows(std::max(row + 1, NRows()));
+    block_changed_ = true;
+  }
+
+  std::vector<unsigned char> packed_buffer_;
```

Comment:
Maybe you could document that this buffer is only used temporarily for (un)packing booleans, and that defining it as a member avoids memory allocations.